### PR TITLE
ci: run integration test on github actions

### PR DESCRIPTION
**Pull request by subhamkrai**
**Description of your changes:**

as we are planning to move toward Github actions
for integration test. This commit runs ceph 
suites using Github action.

Signed-off-by: subhamkrai <srai@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Review by leseb — COMMENTED** [review]
FYI I've made changes in the validation script to only validate certain components in https://github.com/rook/rook/pull/6474

You should probably wait a bit to get it merged and use it

File: .github/workflows/integration-test.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+name: Integration test
+on: [pull_request]
+
+jobs:
+  test:
```
Comment (leseb) [inline .github/workflows/integration-test.yaml:5]:
```suggestion
  CephSmokeSuite:
```

**Review by leseb — CHANGES_REQUESTED** [review]
Looks good so far! Can you try adding more `job`?

**Comment by subhamkrai** [comment]
> Looks good so far! Can you try adding more `job`?

Okay, I'll try adding more `Ceph` tests.

File: .github/workflows/integration-test.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        sudo lsblk
+
+    - name: build rook
+      run: |
+        # set VERSION to a dummy value since Jenkins normally sets it for us. Do this to make Helm happy and not fail with "Error: Invalid Semantic Version"
+        GOPATH=$(go env GOPATH) make clean && make -j$nproc IMAGES='ceph' VERSION=0 build
+        docker images
+        docker tag $(docker images|awk '/build-/ {print $1}') rook/ceph:master
+
+    - name: TestCephSmokeSuite
```
Comment (travisn) [inline .github/workflows/integration-test.yaml:43]: If there is a failure, how can we see the operator or other pod logs?

**Review by travisn — COMMENTED** [review]
> > Looks good so far! Can you try adding more `job`?
> 
> Okay, I'll try adding more `Ceph` tests.

Are we going to have a separate github action for each test suite? If we run multiple suites in the same rule it is going to take too long.

File: .github/workflows/integration-test.yaml (hunk 1 above)
Comment (subhamkrai) [reply .github/workflows/integration-test.yaml]: Done.

File: .github/workflows/integration-test.yaml (hunk 2 above)
Comment (subhamkrai) [reply .github/workflows/integration-test.yaml:43]: Now, we can see logs of any pod other but for now, added only operator pod log.

File: .github/workflows/integration-test.yaml [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+       export TEST_SCRATCH_DEVICE=$(sudo lsblk --paths|awk '/14G/ {print $1}'| head -1)1
+       sed -i "s|deviceFilter:  ''|deviceFilter: $(lsblk|awk '/14G/ {print $1}'| head -1)|g" tests/framework/installer/ceph_manifests.go
+       go test -v -timeout 1800s -run CephMultiClusterDeploySuite github.com/rook/rook/tests/integration
+
+    - name: The job has failed
+      if: ${{ failure() }}
+      run: | 
+       cd /home/runner/work/rook/rook/tests/integration/_output/tests/
+       ls
+       FILE=$(find . ! -name '*_previous_*' -name 'TestCephMultiClusterDeploySuite_localhost_mrc-n1-system_rook-ceph-operator-*.log')
```
Comment (leseb) [inline .github/workflows/integration-test.yaml:284]: We should probably write a collector script at some point.

Comment (subhamkrai) [reply .github/workflows/integration-test.yaml:284]: Agreed. I'll do it once all the tests will pass.

**Comment by BlaineEXE** [comment]
Should we, with this PR, also set it so that the test added to GH actions don't run in Jenkins by default?

**Comment by travisn** [comment]
> Should we, with this PR, also set it so that the test added to GH actions don't run in Jenkins by default?

Yes, there was a separate discussion that we want to skip the Jenkins integration tests if `[test ceph]` (or `ceph:` in the PR title) and **not** `[test full]`

**Comment by BlaineEXE** [comment]
> > Should we, with this PR, also set it so that the test added to GH actions don't run in Jenkins by default?
> 
> Yes, there was a separate discussion that we want to skip the Jenkins integration tests if `[test ceph]` (or `ceph:` in the PR title) and **not** `[test full]`

Sounds like that's for a different PR.

**Comment by travisn** [comment]
@subhamkrai Some comments on the test failures:

The wrong version of the snapshotter is being installed. We should be installing the v3.0.0 snapshotter
```
        	            	Failed to run: kubectl [create -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v2.1.0/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml]
```
I see this in the test code:
```
snapshotterVersion = "v2.1.0"
```
But it's expected as seen [here](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/operator.yaml#L52).

In the `TestBlockStoreOnRookInstalledViaHelm`, it seems the CSI driver isn't mounting the volume. I wonder if there is some related issue with the CSI driver from the first issue above. But we need CSI logs to understand more, I only see the operator log.

In the MultiCluster suite, the operator log shows that the mons are not starting. The mon logs may be needed, it's difficult to know the issue without them.

**Comment by subhamkrai** [comment]
@travisn I have added logs of all the files.
And about snapshotter version, if we see [here](https://github.com/rook/rook/blob/master/tests/framework/utils/snapshot.go#L28) we use `"v2.1.0"` and it is passing in others PR. Also, changed the version to `"v3.0.0"` but still failing.

**Comment by travisn** [comment]
The 3.0.0 snapshotter doesn't exist at that same URL, if @Madhu-1 could take a look at that test.

For the mon not starting, looks like it's not finding a local volume. Is the test calling `localPathPV.sh`? Looks like that script sets up the local PV.
```
27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              MountVolume.NewMounter initialization failed for volume "local-vol1" : path "/var/lib/rook/rook-integration-test/mon1" does not exist
27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              Unable to attach or mount volumes: unmounted volumes=[ceph-daemon-data], unattached volumes=[rook-ceph-mons-keyring rook-ceph-log rook-ceph-crash ceph-daemon-data default-token-gcm7k rook-config-override]: timed out waiting for the condition
```

**Comment by subhamkrai** [comment]
> For the mon not starting, looks like it's not finding a local volume. Is the test calling `localPathPV.sh`? Looks like that script sets up the local PV.
> 
> ```
> 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> MountVolume.NewMounter initialization failed for volume "local-vol1" : path "/var/lib/rook/rook-integration-test/mon1" does not exist
> 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> Unable to attach or mount volumes: unmounted volumes=[ceph-daemon-data], unattached volumes=[rook-ceph-mons-keyring rook-ceph-log rook-ceph-crash ceph-daemon-data default-token-gcm7k rook-config-override]: 
> timed out waiting for the condition
> ```

Yes. it is calling that script.

**Comment by sp98** [comment]
> The 3.0.0 snapshotter doesn't exist at that same URL, if @Madhu-1 could take a look at that test.
> 
> For the mon not starting, looks like it's not finding a local volume. Is the test calling `localPathPV.sh`? Looks like that script sets up the local PV.
> 
> ```
> 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> MountVolume.NewMounter initialization failed for volume "local-vol1" : path "/var/lib/rook/rook-integration-test/mon1" does not exist
> 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> Unable to attach or mount volumes: unmounted volumes=[ceph-daemon-data], unattached volumes=[rook-ceph-mons-keyring rook-ceph-log rook-ceph-crash ceph-daemon-data default-token-gcm7k rook-config-override]: 
> timed out waiting for the condition
> ```

Could be an issue with the [minikube](https://github.com/kubernetes/minikube/issues/4634). It wants the path to be created inside the minikube vm.  Workaround of [creating  the directory](https://github.com/kubernetes/minikube/issues/4634#issuecomment-523070730) - `/var/lib/rook/rook-integration-test/mon1  .. /mon2 , ../mon3` in minikube  fixes the issue with mons.

**Comment by subhamkrai** [comment]
> > For the mon not starting, looks like it's not finding a local volume. Is the test calling `localPathPV.sh`? Looks like that script sets up the local PV.
> > ```
> > 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> > MountVolume.NewMounter initialization failed for volume "local-vol1" : path "/var/lib/rook/rook-integration-test/mon1" does not exist
> > 27s         Warning   FailedMount            pod/rook-ceph-mon-a-5446448f78-hn7qp                                              
> > Unable to attach or mount volumes: unmounted volumes=[ceph-daemon-data], unattached volumes=[rook-ceph-mons-keyring rook-ceph-log rook-ceph-crash ceph-daemon-data default-token-gcm7k rook-config-override]: 
> > timed out waiting for the condition
> > ```
> 
> Could be an issue with the [minikube](https://github.com/kubernetes/minikube/issues/4634). It wants the path to be created inside the minikube vm. Workaround of [creating the directory](https://github.com/kubernetes/minikube/issues/4634#issuecomment-523070730) - `/var/lib/rook/rook-integration-test/mon1 .. /mon2 , ../mon3` in minikube fixes the issue with mons.

@sp98 Thanks for taking a look. It seems to work now.

File: .github/workflows/integration-test.yaml [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+        minikube version: 'v1.13.1'
+        kubernetes version: 'v1.19.2'
+        driver: docker
+        start args: --memory 6g --cpus=2
+
+    - name: check k8s cluster status
+      run: |
+        kubectl cluster-info
+        kubectl get pods -n kube-system
+        minikube ssh -- sudo mkdir -p /var/lib/rook/rook-integration-test/mon1 /var/lib/rook/rook-integration-test/mon2 /var/lib/rook/rook-integration-test/mon3
```
Comment (leseb) [inline .github/workflows/integration-test.yaml:256]: You don't need that if you use the `bare` provider. Why do you use `docker` all of the sudden?

Comment (subhamkrai) [reply .github/workflows/integration-test.yaml:256]: Yeah, I'm using `docker` only for `TestCephMultiClusterDeploySuite ` this test was failing earlier with some reason...I don't remember the error but while running the script `localPathPV.sh`. I'll try again with `bare` and see if it's working.

Comment (leseb) [reply .github/workflows/integration-test.yaml:256]: You might as well run the `localPathPV.sh` before it helps. bare should work as all the other canary scenario use it successfully.

**Comment by subhamkrai** [comment]
> The 3.0.0 snapshotter doesn't exist at that same URL, if @Madhu-1 could take a look at that test.
#6642

**Comment by leseb** [comment]
> > The 3.0.0 snapshotter doesn't exist at that same URL, if @Madhu-1 could take a look at that test.
> > #6642

Ready for a rebase!

File: .github/workflows/integration-test.yaml [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+      run: |
+       export TEST_SCRATCH_DEVICE=$(sudo lsblk --paths|awk '/14G/ {print $1}'| head -1)1
+       sed -i "s|deviceFilter:  ''|deviceFilter: $(lsblk|awk '/14G/ {print $1}'| head -1)|g" tests/framework/installer/ceph_manifests.go
+       go test -v -timeout 1800s -run CephMultiClusterDeploySuite github.com/rook/rook/tests/integration
+
+    - name: Artifact
+      uses: actions/upload-artifact@v2
+      if: failure()
+      with:
+        name: my-artifact
```
Comment (leseb) [inline .github/workflows/integration-test.yaml:307]:
```suggestion
        name: ceph-multi-cluster-deploy-suite-artifact
```

**Review by leseb — APPROVED** [review]
Apply the suggestion to all the jobs

File: .github/workflows/integration-test.yaml (hunk 5 above)
Comment (subhamkrai) [reply .github/workflows/integration-test.yaml]: Done.

**Review by travisn — APPROVED** [review]
Next let's consider adding the CephMgrSuite, but keep it optional since it runs against ceph master.

File: .github/workflows/integration-test.yaml [hunk 6]
```diff
@@ -240,6 +244,8 @@ jobs:
       run: |
        tests/scripts/minikube.sh helm
        tests/scripts/helm.sh up
+       sed -i "s|deviceFilter:  ''|deviceFilter: $(lsblk|awk '/14G/ {print $1}'| head -1)|g" tests/framework/installer/ceph_manifests.go
+       sed -i "s|for i := 0; i < 30; i++ {|for i := 0; i < 60; i++ {|g" tests/framework/utils/k8s_helper.go
```
Comment (leseb) [inline .github/workflows/integration-test.yaml:248]: why not just edit `tests/framework/utils/k8s_helper.go` directly?

**Review by leseb — CHANGES_REQUESTED** [review]

File: .github/workflows/integration-test.yaml (hunk 6 above)
Comment (subhamkrai) [reply .github/workflows/integration-test.yaml]: Done.

**Review by leseb — APPROVED** [review]

**Comment by subhamkrai** [comment]
> Next let's consider adding the CephMgrSuite, but keep it optional since it runs against ceph master.

#6669

**Pull request merged**